### PR TITLE
Homework CMake / Tests

### PR DESCRIPTION
**Pull request by Kasperice**
CMake and tests only
Authors: @Kasperice, @g00fer, @kasiasieradzka, @Pawellos, @DamiXioR

### CODE REVIEW

**Comment by Kasperice** [comment]
@ziobron when I added main.yml when we were working on the project I was receiving some emails (for example: Run Failed...), is it possible to see results of build and test in PR? Or something else should be done to see the result in "Checks"?

File: shm/ship.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+uint32_t Ship::getCrew() const {
+    return crew_;
+}
+
+uint32_t Ship::getMaxCrew() const {
+    return maxCrew_;
+}
+
+uint32_t Ship::getCapacity() const {
```
Comment (kyrtaq) [inline shm/ship.cpp:61]: getters should be implemented in hpp

**Review by kyrtaq — CHANGES_REQUESTED** [review]

File: shm/ship.cpp [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        if (cargo.get()->getName() == element->getName()) {
+            *element += cargo.get()->getAmount();
+            return;
+        }
+    }
+    cargo_.emplace_back(cargo);
+    time_->registerObserver(cargo_.back().get());
+}
+
+void Ship::unload(std::shared_ptr<Cargo> cargo, uint32_t amount) {
```
Comment (kyrtaq) [inline shm/ship.cpp:87]: are you sure about this amount var? it isnt specified in the uml

File: shm/ship.cpp (hunk 1 above)
Comment (Kasperice) [reply shm/ship.cpp:61]: This is only cmake/tests PR

File: shm/ship.cpp (hunk 2 above)
Comment (Kasperice) [reply shm/ship.cpp:87]: This is only cmake/tests PR. BTW we don't have to follow the UML. We decided to use amount (but discussion can be continued when we will create PR with SHM#2)

Comment (kyrtaq) [reply shm/ship.cpp:87]: hmm okay

File: shm/ship.cpp (hunk 1 above)
Comment (kyrtaq) [reply shm/ship.cpp:61]: ah okay sorry

Comment (Kasperice) [reply shm/ship.cpp:61]: and we are defining all getters (even if they are short and easy) and constructors in cpp file, @nauka-programowania-MA was talking about it somewhere

File: shm/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+# dependencies automatically when using CMake 2.8.11 or
+# later. Otherwise we have to add them here ourselves.
+if (CMAKE_VERSION VERSION_LESS 2.8.11)
+  include_directories("${gtest_SOURCE_DIR}/include")
+endif()
+
+#Prepare and compile SHM project
+project(SHM)
+
+set(SRC cargo.cpp coordinates.cpp island.cpp map.cpp ship.cpp player.cpp store.cpp time.cpp observerTime.cpp subjectTime.cpp alcohol.cpp item.cpp fruit.cpp)
```
Comment (ziobron) [inline shm/CMakeLists.txt]: it's better to have a multiline list. It's easier to see on diff which files are added or removed.

File: shm/main.cpp [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    // std::shared_ptr<Cargo> ptrCargo1 = std::make_shared<Cargo>(Chocolate);
+    // std::shared_ptr<Cargo> ptrCargo2 = std::make_shared<Cargo>(IceCream);
+
+    // pawellos.getShip()->load(ptrCargo1);
+    // pawellos.getShip()->load(ptrCargo2);
+    std::cout << "---------Before buy------ \n\n";
+    pawellos.getShip()->printCargo();
+    std::cout << "Money of player: " << pawellos.getMoney() << '\n';
+    std::cout << "Available place on ship: " << pawellos.getAvailableSpace() << '\n';
+    Store biedronka;
```
Comment (ziobron) [inline shm/main.cpp]: 😄

File: shm/observerTime.cpp [hunk 5]
```diff
@@ -0,0 +1 @@
+#include "observerTime.hpp"
```
Comment (ziobron) [inline shm/observerTime.cpp]: file not needed. Can be removed. Hpp is enough for full abstract classes.

File: shm/subjectTime.cpp [hunk 6]
```diff
@@ -0,0 +1 @@
+#include "subjectTime.hpp"
```
Comment (ziobron) [inline shm/subjectTime.cpp]: File not needed.

File: shm/test.cpp [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+
+TEST_F(SHM_TestFixture, ShouldAlwaysPass) {
+
+    EXPECT_TRUE(true);
+}
+
+TEST_F(SHM_TestFixture, ShouldReturnNullPtrWhenAskedForIslandOutOfRange) {
+
+    Coordinates location(MAP_RND_DISTRIBUTION_MAX + 10, MAP_RND_DISTRIBUTION_MAX + 10);
```
Comment (ziobron) [inline shm/test.cpp]: testMap doesn't use this location variable, does it?

File: shm/ut/alcohol_test.cpp [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    EXPECT_EQ("Alcohol " + NAME_1, alko_1.getName());
+}
+
+TEST_F(SHM_TestFixtureAlco, ShouldReturnInitializedAmount) {
+
+    EXPECT_EQ(AMOUNT_1, alko_1.getAmount());
+}
+
+TEST_F(SHM_TestFixtureAlco, ShouldReturnInitializedBasePrice) {
+
```
Comment (ziobron) [inline shm/ut/alcohol_test.cpp]: this empty line in every test is not uniform with the rest of your code.

File: shm/ut/alcohol_test.cpp [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    Alcohol alko_2(AMOUNT_1, NAME_1, BASEPRICE_1, POWER_1) ;
+    Alcohol alko_3(0, "wtf", 0, 0) ;
+
+    if(alko_1 == alko_2){
+        EXPECT_EQ(true, true);   
+    }
+    if(alko_1 == alko_3){
+        EXPECT_EQ(false, true);
+    }
```
Comment (ziobron) [inline shm/ut/alcohol_test.cpp]: You want something like this:
```cpp
EXPECT_EQ(alko_1, alko_2);
EXPECT_NE(alko_1, alko_3);
```

File: shm/ut/coordinates_test.cpp [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+    Coordinates coord_2(100, 100) ;
+    Coordinates coord_3(0, 0) ;
+
+    if(coord_1 == coord_2){
+        EXPECT_EQ(true, true);   
+    }
+    if(coord_1 == coord_3){
+        EXPECT_EQ(false, true);
+    }
```
Comment (ziobron) [inline shm/ut/coordinates_test.cpp]:
```suggestion
    EXPECT_EQ(coord_1, coord_2);   
    EXPECT_NE(coord_1, coord_3);
```

**Review by ziobron — APPROVED** [review]
CMake looks good. I can see that your CI works on your fork. Can you try to do `git pull --rebase coders master` to have CI working on PR as well ;)

File: shm/test.cpp (hunk 7 above)
Comment (Kasperice) [reply shm/test.cpp]: testMap is not using this variable. testMap has only 10 islands with coordinates between MAP_RND_DISTRIBUTION_MIN and MAP_RND_DISTRIBUTION_MAX. We want to check only if nullptr is returned if island is not found in map

File: shm/ut/alcohol_test.cpp (hunk 9 above)
Comment (Kasperice) [reply shm/ut/alcohol_test.cpp]: For NE we will have to implement operator!=()?

File: shm/ut/coordinates_test.cpp (hunk 10 above)
Comment (Kasperice) [reply shm/ut/coordinates_test.cpp]: For NE we will have to implement operator!=()?

**Comment by Kasperice** [comment]
@ziobron I did the rebase and implement some changes, could you look at this

**Comment by Kasperice** [comment]
@ziobron I have also two comments regarding "checker":
1) Check ".github/workflows/main.yml". Should the .github directory be inside shm directory? I think it should be in main directory, so check should verify "../.github/workflows/main.yml"
2) Check "binary with tests exist". We prepared test binary in one line "add_executable(${PROJECT_NAME}-ut test.cpp ${TEST_SRC})" But check require us to divide it into two lines, is it necessary or can we leave it as it is?

**Comment by Kasperice** [comment]
@ziobron i pushed some changes to our shm2 branch (I forgot that this PR is opened) and now Checks are failing in "checker" with error: ```Err:1 http://ppa.launchpad.net/ondrej/php/ubuntu bionic/main amd64 pcregrep amd64 2:8.43-1+ubuntu18.04.1+deb.sury.org+1
  404  Not Found [IP: 91.189.95.83 80]```
Do you know what caused this error?

The last working commit was d1812a1. We did not change anything in main.yml file since that time. As I found in Internet it can be better to use sudo apt**-get** install (...) in those .yml scripts, but I tried also this and it did not resolve the problem.

**Comment by Kasperice** [comment]
@ziobron I managed to correct the "checker" job (I had to add one line `sudo apt-get update`) I hope you will not be mad that I have changed your code a bit

File: shm/cargo.hpp [hunk 11]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include "subjectTime.hpp"
+#include "time.hpp"
```
Comment (WBigaj) [inline shm/cargo.hpp:6]: I think there should be a new line between standard library and headers include.

File: shm/game.hpp [hunk 12]
```diff
@@ -0,0 +1,39 @@
+#pragma once
+#include <memory>
+#include <string>
```
Comment (WBigaj) [inline shm/game.hpp:3]: I would put newline between pragma and includes.

File: shm/game.hpp [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+private:
+    uint32_t money_;
+    uint32_t days_;
+    uint32_t target_;
+    Time* time_;
+
+    uint32_t timeElapsed = 0;
+
+    std::shared_ptr<Ship> ship_ = std::make_shared<Ship>(200, 50, 30, "black pearl", 1, time_);
+    std::shared_ptr<Player> player_ = std::make_shared<Player>(ship_, 300, 80);
```
Comment (WBigaj) [inline shm/game.hpp:35]: What are those magic numbers. Isn't better to make them more clear by making constexpr like startShipSize.

**Review by WBigaj — COMMENTED** [review]
I have checked your code and gave my feedback.

File: shm/game.hpp (hunk 13 above)
Comment (Kasperice) [reply shm/game.hpp:35]: If you have some comments please leave them below right PR, this one is connected to CMake/tests

**Comment by ziobron** [comment]
> @ziobron I have also two comments regarding "checker":
> 
> 1. Check ".github/workflows/main.yml". Should the .github directory be inside shm directory? I think it should be in main directory, so check should verify "../.github/workflows/main.yml"
> 2. Check "binary with tests exist". We prepared test binary in one line "add_executable(${PROJECT_NAME}-ut test.cpp ${TEST_SRC})" But check require us to divide it into two lines, is it necessary or can we leave it as it is?

1. Yes, my check is incorrect. 
2. Probably I missed your case and regex require improvement. Don't worry about it, I look at the code anyway ;)

**Review by ziobron — APPROVED** [review]

File: shm/ut/alcohol_test.cpp (hunk 9 above)
Comment (ziobron) [reply shm/ut/alcohol_test.cpp]: I think so. You can do it by negating the result of the operator==

File: shm/cargo.hpp (hunk 11 above)
Comment (ziobron) [reply shm/cargo.hpp:6]: it's a matter of the group internal convention and you should not comment the coding style in group projects.

**Comment by ziobron** [comment]
10 x 2XP for each point in cmake homework (2767)
2 x 2XP for timely delivery (1119)

{
  "requirement_ids": [2767, 2767, 2767, 2767, 2767, 2767, 2767, 2767, 2767, 2767, 1119, 1119],
  "users": ["Kasperice", "g00fer", "kasiasieradzka", "Pawellos", "DamiXioR"]
}

**Pull request closed without merging**